### PR TITLE
Flesh out the server

### LINKED ISSUE #7: Prompts endpoints
- [x] random prompt
- [x] random prompt by genre
- [x] specific prompt
- [x] all prompts

### LINKED ISSUE #8: Create flag to determine if story is finished/publishable on `story` table

### LINKED ISSUE #9: Author endpoints
- [x] get all authors
- [x] get specific authors
- [x] edit author bios
- [x] edit author passwords
- [x] edit author email address
- [x] post new author 
  - [x] error handling for username or email already taken
- [x] post login author

### LINKED ISSUE #12: Story data type
The story array is not coming back as an array - assuming that postman is running .json (which I blieve it is)

image

### LINKED ISSUE #14: Posting stories
- [x] Posting new stories
- [x] Posting additions to new stories
  - [x] completing stories

### PR DESCRIPTION
**Pull request by GreysonElkins**
#### What's this PR do?
* introduces Helper classes with static methods for the server
* installs a package to allow the server to handle static methods
* provides temporary seeds for authors / stories, I suggest removing or replacing these before deployment. 
* _Author endpoints_ (password protected): **get** single or all, **post** new, **post** login, **patch** email password and/or bio
* _Prompt endpoints_: **get**, all, any random single, or random by genre
* _Stories endpoints_: **get** all, single by id, **post** new, **patch** update old
#### Where should the reviewer start?
* The **stories post endpoint** for posting new stories was updated last minute to reflect new columns / data-types. Some testing here especially, and fleshing out error responses would be immensly helpful!
* probably are going to need to run `npm install --save-dev @babel/plugin-proposal-class-properties`
* familiarize yourself with the endpoints, what data they require, and what data they don't allow. 
#### How should this be manually tested?
Get this server launched and play around with some data in Postman!
#### Any background context you want to provide?
`StoryHelper.checkForProblems` may be overkill - once I got through `StoryHelper.updateStory` I realized it would only be updating things that were allowed to be updated anyway. 
#### What are the relevant tickets?
#6, close #7, close #8, close #9, close #12, close #14 
#### Screenshots (if appropriate)
giphy
#### Questions:
* any good error handling oppurtunites, or other endpoints y'all would like to see?

### CODE REVIEW

File: server-helpers.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    }
+  }
+
+  static findSpecificPrompt = (id) => {
+    return knex("prompts")
+      .groupBy("id")
+      .select()
+      .having("id", "=", id)
+  }
+}
```
Comment (nickhartdev) [inline server-helpers.js:108]: I added an issue for this, but I'm guessing something in here (guessing the `Math.random()`) is causing a blank response is being returned on occasion.

File: server-helpers.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+      } else if (field === 'prompt' || field === 'title' || field === 'id') {
+        problems.push(`${fields}s are not allowed to be updated`)
+      } else if (field === 'is_complete' && typeof story.is_complete !== "boolean") {
+        problems.push('Stories being complete should be true or false.')
+      } else if (field === 'created_at' || field === 'updated_at') {
+        problems.push('You can not manually update publishing dates')
+      }
+    })
+    return problems
+  }
```
Comment (nickhartdev) [inline server-helpers.js:130]: I really appreciate how thorough you got with this, this'll save us a lot in the long run

**Review by nickhartdev — APPROVED** [review]
Really solid stuff! Left a comment for the one piece of feedback I'd give, I appreciate the thoroughness of the error handling built in here.

**Review by carlymclift — APPROVED** [review]
Thanks for working so hard on this

File: db/migrations/20200913122453_story_authors+datatype.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+
+exports.up = function(knex) {
+  return knex.schema.table('stories', table=> {
+    table.specificType('contributions', 'text ARRAY');
```
Comment (Abdeboskey) [inline db/migrations/20200913122453_story_authors+datatype.js:4]: Just a little thing, but adding a space before the thicc arrow would be gr8.

File: db/migrations/20200913124040_story_completion.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+
+exports.up = function(knex) {
+  return knex.schema.table("stories", (table) => {
+    table.boolean('is_complete').defaultTo(0)
+  });
```
Comment (Abdeboskey) [inline db/migrations/20200913124040_story_completion.js:5]: does `0` here default to false? like Binary?

File: db/seeds/dev/authors_seed.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            name: 'Jacob Jinger Heimer Schmidt',
+            email: 'comegetyourpancakes@ihop.com',
+            created_at: "2020-05-11T00:25:34.315Z",
+            updated_at: "2020-06-11T00:25:34.315Z",
+            password: 'querty',
+            bio: 'Some things never change. I am a magnet.'
+          }
+        ]);
+      });
+};
```
Comment (Abdeboskey) [inline db/seeds/dev/authors_seed.js:53]: This is some high quality mock data right here.

File: db/seeds/dev/stories_seed.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            his front two legs. He screamed in agony because his worst fears had come true, he lost his marbles.`,
+              `Having lost his marbles he could only rely upon his intuition, a rare energy among the centaurs 
+            a space based power within his mind that was devoid of thoughts and rich with imagination.`
+          ],
+          contributors: [5, 4, 3],
+          is_complete: false,
+        }
+      ]);
+    });
+};
```
Comment (Abdeboskey) [inline db/seeds/dev/stories_seed.js:115]: What an awesome way to preserve the OG Corpses that we made together way back when. I would vote that we keep these in here for deployment.

File: server-helpers.js [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+      created_at: created_at,
+      updated_at: updated_at,
+    };
+  }
+
+  static checkLoginInfo = (info) => {
+    const infoProvided = Object.keys(info)
+    const result = { hasEnoughInfo: false, message: '', userType: ''}
+    if (!infoProvided.includes("password")) {
+      result.message = 'You need to provide a password'
```
Comment (Abdeboskey) [inline server-helpers.js:42]: Won't this always include a key of password even if the value wasn't submitted? Would something like `if (!infoProvided[password])` make more sense? Or maybe I'm misunderstanding?

File: server-helpers.js [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+      if (userAccount[0].password === attemptedPassword) {
+        return response.status(200).json(this.makeSecureUserResponse(userAccount[0]))
+      } else {
+        return response.status(422).json('An incorrect password was provided!')
+      }
+    } catch (error) {
+      console.error(error.message)
+    }
+  }
+
```
Comment (Abdeboskey) [inline server-helpers.js:63]: I dig this.

**Review by Abdeboskey — APPROVED** [review]
This is a lot of work GreySON, thank you for making this happen!

**Pull request merged**